Struct to Stringer var assignment panics

The following program `sample.go` triggers a panic:

```go
package main

import (
	"fmt"
)

type foo struct {
	bar string
}

func (f foo) String() string {
	return f.bar
}

func main() {
	var f fmt.Stringer = foo{bar: "bar"}
	fmt.Println(f)
}
```

Expected result:
```console
$ go run ./sample.go
bar
```

Got:
```console
$ yaegi ./sample.go
./sample.go:16:8: panic
reflect.Set: value of type struct { Xbar string } is not assignable to type fmt.Stringer
goroutine 1 [running]:
runtime/debug.Stack(0x1, 0xc000148400, 0x40)
        /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/debug/stack.go:24 +0x9d
github.com/containous/yaegi/interp.(*Interpreter).Eval.func1(0xc0001a7e00)
        /Users/berng/go/src/github.com/containous/yaegi/interp/interp.go:320 +0xbc
panic(0x1896ea0, 0xc000192a70)
        /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/panic.go:969 +0x166
github.com/containous/yaegi/interp.runCfg.func1(0xc0001357c0, 0xc000300a00)
        /Users/berng/go/src/github.com/containous/yaegi/interp/run.go:113 +0x20c
panic(0x1896ea0, 0xc000192a70)
        /usr/local/Cellar/go/1.14.2_1/libexec/src/runtime/panic.go:969 +0x166
reflect.Value.assignTo(0xc0001207e0, 0xc000192a00, 0x199, 0x19c9d58, 0xb, 0x18edb60, 0xc000192a40, 0xc0002da8a0, 0xc000192860, 0xc0001425a0)
        /usr/local/Cellar/go/1.14.2_1/libexec/src/reflect/value.go:2403 +0x426
reflect.Value.Set(0x18edb60, 0xc000192a40, 0x194, 0xc0001207e0, 0xc000192a00, 0x199)
        /usr/local/Cellar/go/1.14.2_1/libexec/src/reflect/value.go:1532 +0xbd
github.com/containous/yaegi/interp.doCompositeSparse.func1(0xc0001357c0, 0xc000043f80)
        /Users/berng/go/src/github.com/containous/yaegi/interp/run.go:1871 +0x264
github.com/containous/yaegi/interp.runCfg(0xc000300a00, 0xc0001357c0)
        /Users/berng/go/src/github.com/containous/yaegi/interp/run.go:119 +0x6d
github.com/containous/yaegi/interp.(*Interpreter).run(0xc000001980, 0xc000300000, 0xc000135720)
        /Users/berng/go/src/github.com/containous/yaegi/interp/run.go:97 +0x239
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000001980, 0xc0002fe000, 0xb8, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/berng/go/src/github.com/containous/yaegi/interp/interp.go:388 +0x41f
main.main()
        /Users/berng/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:139 +0x554
```

--- a/_test/interface38.go
+++ b/_test/interface38.go
@@ -1,0 +1,19 @@
+package main
+
+import "fmt"
+
+type foo struct {
+	bar string
+}
+
+func (f foo) String() string {
+	return "Hello from " + f.bar
+}
+
+func main() {
+	var f fmt.Stringer = foo{bar: "bar"}
+	fmt.Println(f)
+}
+
+// Output:
+// Hello from bar

--- a/_test/interface39.go
+++ b/_test/interface39.go
@@ -1,0 +1,19 @@
+package main
+
+import "fmt"
+
+type foo struct {
+	bar string
+}
+
+func (f *foo) String() string {
+	return "Hello from " + f.bar
+}
+
+func main() {
+	var f fmt.Stringer = &foo{bar: "bar"}
+	fmt.Println(f)
+}
+
+// Output:
+// Hello from bar

--- a/_test/interface40.go
+++ b/_test/interface40.go
@@ -10,12 +10,12 @@ func (f foo) String() string {
 	return "Hello from " + f.bar
 }
 
-func NewFoo(s string) fmt.Stringer {
+func Foo(s string) fmt.Stringer {
 	return foo{s}
 }
 
 func main() {
-	f := NewFoo("bar")
+	f := Foo("bar")
 	fmt.Println(f)
 }
 

--- a/_test/interface40.go
+++ b/_test/interface40.go
@@ -1,0 +1,23 @@
+package main
+
+import "fmt"
+
+type foo struct {
+	bar string
+}
+
+func (f foo) String() string {
+	return "Hello from " + f.bar
+}
+
+func NewFoo(s string) fmt.Stringer {
+	return foo{s}
+}
+
+func main() {
+	f := NewFoo("bar")
+	fmt.Println(f)
+}
+
+// Output:
+// Hello from bar

--- a/_test/interface41.go
+++ b/_test/interface41.go
@@ -10,12 +10,12 @@ func (f *foo) String() string {
 	return "Hello from " + f.bar
 }
 
-func NewFoo(s string) fmt.Stringer {
+func Foo(s string) fmt.Stringer {
 	return &foo{s}
 }
 
 func main() {
-	f := NewFoo("bar")
+	f := Foo("bar")
 	fmt.Println(f)
 }
 

--- a/_test/interface41.go
+++ b/_test/interface41.go
@@ -1,0 +1,23 @@
+package main
+
+import "fmt"
+
+type foo struct {
+	bar string
+}
+
+func (f *foo) String() string {
+	return "Hello from " + f.bar
+}
+
+func NewFoo(s string) fmt.Stringer {
+	return &foo{s}
+}
+
+func main() {
+	f := NewFoo("bar")
+	fmt.Println(f)
+}
+
+// Output:
+// Hello from bar

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -507,9 +507,13 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					src.findex = dest.findex // Set recv address to LHS
 					dest.typ = src.typ
 				case n.action == aAssign && src.action == aCompositeLit:
-					n.gen = nop
-					src.findex = dest.findex
-					src.level = level
+					if !(dest.typ.cat == valueT && dest.typ.rtype.Kind() == reflect.Interface) {
+						// Output literal composite to destination (avoiding assign operation)
+						// only if there is no need to wrap to an interface.
+						n.gen = nop
+						src.findex = dest.findex
+						src.level = level
+					}
 				case src.kind == basicLit:
 					// TODO: perform constant folding and propagation here.
 					switch {
@@ -884,9 +888,7 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 
 		case compositeLitExpr:
 			wireChild(n)
-			if n.anc.action != aAssign {
-				n.findex = sc.add(n.typ)
-			}
+			n.findex = sc.add(n.typ)
 			// TODO: Check that composite literal expr matches corresponding type
 			n.gen = compositeGenerator(n, sc)
 

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -507,13 +507,15 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					src.findex = dest.findex // Set recv address to LHS
 					dest.typ = src.typ
 				case n.action == aAssign && src.action == aCompositeLit:
-					if !(dest.typ.cat == valueT && dest.typ.rtype.Kind() == reflect.Interface) {
-						// Output literal composite to destination (avoiding assign operation)
-						// only if there is no need to wrap to an interface.
-						n.gen = nop
-						src.findex = dest.findex
-						src.level = level
+					if dest.typ.cat == valueT && dest.typ.rtype.Kind() == reflect.Interface {
+						// No optimisation attempt for assigned binary interface
+						break
 					}
+					// Skip the assign operation entirely, the source frame index is set
+					// to destination index, avoiding extra memory alloc and duplication.
+					n.gen = nop
+					src.findex = dest.findex
+					src.level = level
 				case src.kind == basicLit:
 					// TODO: perform constant folding and propagation here.
 					switch {

--- a/interp/run.go
+++ b/interp/run.go
@@ -1595,6 +1595,12 @@ func _return(n *node) {
 			values[i] = genValue(c)
 		case interfaceT:
 			values[i] = genValueInterface(c)
+		case valueT:
+			if t.rtype.Kind() == reflect.Interface {
+				values[i] = genInterfaceWrapper(c, t.rtype)
+				break
+			}
+			fallthrough
 		default:
 			if c.typ.untyped {
 				values[i] = genValueAs(c, def.typ.ret[i].TypeOf())

